### ISSUE TITLE
Command success or failure ought to feed back into history autosuggestions

### ISSUE THREAD

**Issue by ridiculousfish** [fish-shell/fish-shell#5]
If you mistype a command, we'll happily suggest it again, even if the command failed to run. We ought to feed command success or failure back into history so it won't autosuggest that. But we need a way to distinguish between commands that could not run and commands that ran successfully, yet indicated failure to communicate some other information (such as grep's "not found").

**Comment by faho** [comment]
So, I've taken a look at this, and there are some interesting questions.

We'd have to add the status to the history, and then we can access that in the autosuggestions as well as in the history proper.

And for autosuggestions, I'm not sure just ignoring all suggestions with $status > 0 is correct. Because very frequently we'll execute a command, have it fail because e.g. the command doesn't exist (or some other precondition isn't satisfied), install the command and then try it again. In that situation the previously failed command is *the best choice*.

In fact I don't think a falsy $status really says much about how likely the command is to be the right choice on the whole.

What we could do for suggestions is to ignore candidates with *permanent* issues like syntax errors, but that bites a bit with the suggestion being the first history element, because you often want to recall that to correct it.

What I would add is to not *store* syntactically broken commands on disk instead.

So we'd:

- Skip history suggestions that have syntax errors
- Keep history suggestions that have "transient" errors (command not found, no glob match...)
- Keep the suggestion in the history
- Don't store syntactically faulty commands on disk

**Comment by krobelus** [comment]
fish already does not store syntactically invalid commands like `echo ${}}`, whereas bash happily does.

Detecting semantic errors is quite hard because it's hard to avoid false positives.

The [grml zsh config](https://grml.org/zsh/) includes a nifty feature: if you try to run a command that does not exist, it asks you to pick a command with a similar name. This way it avoids adding commands with typos to the history. I think this would be a nice feature in fish.

```zsh
% eco 123
zsh: correct 'eco' to 'echo' [nyae]? y
123
```

**Comment by faho** [comment]
>fish already does not store syntactically invalid commands like echo ${}}, whereas bash happily does.

That's a good point, we already refuse to run many kinds of invalid commands, so you're forced to fix them.

>it asks you to pick a command with a similar name. This way it avoids adding commands with typos to the history. I think this would be a nice feature in fish.

I **hate** that feature with a fiery passion. I mean what I hate most about it is that awful "[nyae]", but I also hate how it pops up even when I mean something different, or how it frequently guesses wrong and makes me press the correct key to not execute the wrong thing and get me back to my commandline.

**Comment by fregante** [comment]
> very frequently we'll execute a command, have it fail because e.g. the command doesn't exist

I think that's the only situation where the command should be stored, but after that 10-minute the command will still stay cluttering the history forever. Could there be a temporary in-memory history for failed command instead? Once you close the session, it's gone.

**Comment by faho** [comment]
>I don't need git colne to ever be suggested again and there's no easy way to delete it.

The problem is: How does fish know the error was the subcommand was invalid, i.e. the command is always unusable, and how does it tell that from `git checkout $branch` failing because the branch doesn't exist, which will become usable once the branch is created?

I don't see an obvious answer to that, and I'm not sure there even is one.

>Calling $ history --delete means you have to search for the command and then accept the list it provides. Too many keystrokes.

See, that's an argument for making it easier to forget the last command - #5791. A `history forget` command that just deletes the last command would work and you could bind it to a key so it's a single keystroke.

**Comment by fregante** [comment]
> I don't see an obvious answer to that, and I'm not sure there even is one.

Does it need to? It doesn't have to be a default, just an option. As long as I can press ArrowUp to get the last failed command and fix it, I don't need it after that in the vast majority of cases.

**Comment by faho** [comment]
>Does it need to? 

Well, yes, because the `git checkout` failure is *likely* to be the correct suggestion because you'll run `git branch` to make it work.

So it's

```fish
> git checkout foo
# error: no branch foo
> git branch foo
> git checkout foo # <- this is very likely!
```

vs

```fish
> git colne # <- this will never be needed again
```

>It doesn't have to be a default, just an option. 

We try not to do that here.

**Comment by fregante** [comment]
> Does it need to?

What I mean by that is that _I_ want to make that decision. Whether it affects your described situation is something I'm willing to put up with.

Also that situation is already covered by the _session-only history_ that I suggested.

**Comment by faho** [comment]
>Is there a command/config you could suggest that automatically drops the command from history when it fails?

Probably something like

```fish
function drop-failures --on-event fish_postexec
    if test $status -ne 0
        history delete $history[1]
    end
end
```

Haven't tested it.

**Comment by faho** [comment]
Okay, I'm rejecting this, because I don't see a way to differentiate between a failure that you correct and then rerun, and a failure that wasn't usable (e.g. a broken command invocation).

**Comment by boxed** [comment]
Not storing syntactically invalid commands and commands that point to a file that don't exist seems like it would solve 90% of the problem. 

The problem with that is that syntactically invalid stuff SHOULD be stored so you can fix them later. So my suggestion is to store the time of execution for those types of commands, and then throw them away if they are older than say a day. It's probably useful to store the exit code and timestamp of commands in the history file anyway for other reasons?

Sorry about beating a dead horse :P

**Comment by nedredmond** [comment]
As a professional with poo brain, I'd love to have the _option_ of skipping commands with "transient" errors, because I frequently type `git` when I meant `make` or `py` when I mean `go`, etc, and then those useless commands are stored forever. I'm sure there are others with my issue here.

**Comment by timvahlbrock** [comment]
> Okay, I'm rejecting this, because I don't see a way to differentiate between a failure that you correct and then rerun, and a failure that wasn't usable (e.g. a broken command invocation).

Some time since this was last touched, but wouldn't it be okay if "temporary failures" were ignored? The history is mostly useful for commands that are typed very often, so chances are at some point of time the command will have succeeded. Regarding the example of the missing command: The initial failed invocation wouldn't be added to the history, but if I'm willing to install the command, I will likely use it more than twice and therefore will add it to the history as soon as I run it once successfully.

Not adding failed commands to the history would limit suggestions to all commands that have succeeded at any point of time in the past. The implementation could even go so far to add the command to the history if the command **either** succeeded now **or** at any point in the past, making it the preferred suggestion if typing is started again.

Additionally, failed commands could be added to the history anyway, but ignored by suggestions. For the missing command, that would mean one doesn't get the suggestion, but can tap the upwards arrow key twice (past the installation command) to get the previously attempted command.

Hoping this can be revived.

**Comment by faho** [comment]
> Some time since this was last touched, but wouldn't it be okay if "temporary failures" were ignored?

What *is* a temporary failure? How does the shell know that e.g. `git push origin` failed because the network was down, but `git push fart` failed because you don't have a remote called "fart" and `git pshu origin` failed because you misspelled "push"?

For outright *syntax errors*, we already don't add them to history to begin with. Anything else is potentially a correct command.

**Comment by timvahlbrock** [comment]
Yeah but that's what I'm saying: If we don't add `git push origin` to the history when the network is down is no problem, because we will do it so often anyway, that some time in the past it will have succeeded.

**Comment by boxed** [comment]
I have a third option:

For each command, keep a timestamp. Drop failed commands from the history if they are older than some amount, say 5 days. 

Or another one:

Failed commands could be searchable, but not auto-expanded (or whatever it's called when fish shows commands you can complete with the right arrow key). 

I think we can think outside the box a bit here and not keep our thinking constrained to "keep" or "not keep", as both those options are undesirable.

**Comment by timvahlbrock** [comment]
> I have a third option:
> 
> For each command, keep a timestamp. Drop failed commands from the history if they are older than some amount, say 5 days.

More implementation, more complexity, less predictable behaviour - at least from my point of view. So maybe as an addiotional, opt in feature?

> Or another one:
> 
> Failed commands could be searchable, but not auto-expanded (or whatever it's called when fish shows commands you can complete with the right arrow key).

That's what I meant with:

>Additionally, failed commands could be added to the history anyway, but ignored by suggestions. For the missing command, that would mean one doesn't get the suggestion, but can tap the upwards arrow key twice (past the installation command) to get the previously attempted command.

> I think we can think outside the box a bit here and not keep our thinking constrained to "keep" or "not keep", as both those options are undesirable.

Maybe, but as said above: Just deleting failed commands shouldn't be too much of problem, because frequently used commands (which are the ones suggestions are most useful for!!!) are not affected, because they will at some time have succeeded.

**Comment by nedredmond** [comment]
If we have already stored a command that has run successfully, do we throw it away when it fails? No, we just don't store it _that_ time.

I don't see the need to distinguish transitory failures, because if they _are_ transitory, by definition they may have succeeded in the past and/or will in the future and will be stored.

On the other hand, "broken invocations" will never pass and just pollute the autocomplete history and are an annoying reminder of my twitchy fingers.

**Comment by ZoomRmc** [comment]
#10302 was merged (thanks!) but it's working [before the command is executed](https://github.com/fish-shell/fish-shell/blob/master/doc_src/cmds/fish_should_add_to_history.rst#description). This does not allow deciding if the command should not be saved based on the program output. For this fish needs can distinguish the types of exit codes to decide.

From my experience, even when the program exits with an error I'd like the invocation to be saved, as in most cases it's either a consequence of:
1. Conditions outside of program's control. Such as: web request with network down, file access to deleted/restricted directory, etc.
2. Bug in a program that's being developed.

In both cases the invocation is likely to be useful in the future.

On the other hand, basic typos/mistakes in history make the user repeatedly fail and saving them is counterproductive. There's two cases:
1. Calling an unavailable command (`rsycn`, `vash`).
2. Mistyping/using the wrong command argument.

The first is easy for the shell to distinguish, while for the second there's a need for some contextual information.
To fix this I'm thinking about proposing a convention for using a specific exit code to signal **user** error, as the source of the program error. The likely candidate is error code **22 EINVAL** ("invalid argument"). If supported by both the CLI frameworks and shells it can be adopted in a reasonable amount of time.

I personally started to use exit code 22 as a signal of user error in my programs CLI logic and would like to be able to tell shells to act accordingly.

**TLDR:**
An actionable solution to allow a path for establishing a convention is to provide an option to reject history records based on specific exit codes, suggesting (but not pre-setting) one as a default user setting.

**Comment by timvahlbrock** [comment]
> From my experience, even when the program exits with an error I'd like the invocation to be saved, as in most cases it's either a consequence of:
>
>Conditions outside of program's control. Such as: web request with network down, file access to deleted/restricted directory, etc.
Bug in a program that's being developed.

As written before: anything that fails because something outside of the programs control should not be an issue, because we likely would have saved this exact command sometime before in history, we just wouldn't add it again. This isn't about deleting commands that failed, this is about saving ones that didn't.

**Comment by ZoomRmc** [comment]
> As written before: anything that fails because something outside of the programs control should not be an issue, because we likely would have saved this exact command sometime before in history, we just wouldn't add it again. This isn't about deleting commands that failed, this is about saving ones that didn't.

I'm not sure what's your point here, exactly. The part you quoted is about examples of cases when the commands _should_ be saved.

**Comment by boxed** [comment]
@timvahlbrock sure, that all makes sense, except that if your first try fails because you mistyped the command. You want to be able access that history item at least for an hour or a day or something to try again.

**Comment by timvahlbrock** [comment]
> @timvahlbrock sure, that all makes sense, except that if your first try fails because you mistyped the command. You want to be able access that history item at least for an hour or a day or something to try again. 

Yeah maybe, but I think autocomplete is mainly useful for commands that I use repeatedly. I'm not sure the benefits of being able to autocomplete failing commands that I have never used before outweighs the additional complexity required to implement and configure this. Even if so, remembering failed commands separately would probably make implementing your proposal easier.
